### PR TITLE
mixin and components

### PR DESCRIPTION
**Pull request by jannyHou**
connect to https://github.com/strongloop/loopback-next/issues/580

- Add mixin doc in `Mixin.md`
- Also add a section for javascript/typescript related concepts as mixin's parent section
- Add how a component interact with mixin in `Creating-component.md`

### CODE REVIEW

**Comment by crandmck** [comment]
@jannyHou I'll wait til this is a bit further along to review it OK?  Pls ping me when it's ready for review.

**Comment by jannyHou** [comment]
@crandmck yeah I am still adding stuff on local, thanks!

File: _data/sidebars/lb4_sidebar.yml
Comment (bajtos) [inline _data/sidebars/lb4_sidebar.yml]: Is it worth starting a new top-level section instead of nesting this new "Mixin" page under "Key concepts"? I don't know what's the right answer, just something to consider.

File: pages/en/lb4/Creating-components.md
Comment (bajtos) [inline pages/en/lb4/Creating-components.md]: I think this should not be strictly necessary, I believe TypeScript is able to infer the return type from `Provider` definition and/or the actual value returned by the code of this method. Unless I am mistaken?

File: pages/en/lb4/Creating-components.md
Comment (bajtos) [inline pages/en/lb4/Creating-components.md]: `implements`?

File: pages/en/lb4/Creating-components.md
Comment (bajtos) [inline pages/en/lb4/Creating-components.md]: `implements`?

File: pages/en/lb4/Creating-components.md
Comment (bajtos) [reply pages/en/lb4/Creating-components.md]: ❗️ actually, the code examples in this file are all using JavaScript, not TypeScript - see `js` after three backticks. I think you should revert type-related changes in this file.

File: _data/sidebars/lb4_sidebar.yml
Comment (jannyHou) [reply _data/sidebars/lb4_sidebar.yml]: @bajtos I feel at least now `Mixin` is not a LB core concept, we don't have mixin level context, and don't reply on it to scaffold the app, it is more like a JS/TS strategy to extend class, which make other concepts extendable, like application, controller, as long as it's a class.

there is some context of this discussion, I pinged Raymond about moving `mixin.ts` from `@loopback/repository` to `@loopback/core`, and he suggests we can have a `@loopback/lang` or `@loopback/common` module for JS/TS related stuff.

I created an issue for this idea: https://github.com/strongloop/loopback-next/issues/660

**Comment by jannyHou** [comment]
Regarding new category `Lang`, there are some context in story https://github.com/strongloop/loopback-next/issues/660

File: pages/en/lb4/Lang-concepts.md
Comment (crandmck) [inline pages/en/lb4/Lang-concepts.md]: Language-related concepts

File: _data/sidebars/lb4_sidebar.yml
Comment (crandmck) [reply _data/sidebars/lb4_sidebar.yml]: S/be 
Language-related concepts

File: pages/en/lb4/Creating-components.md
Comment (crandmck) [inline pages/en/lb4/Creating-components.md]: This section has many grammatical and stylistic errors.  Rather than go through them laboriously here, I'd prefer to either edit the text in thebranch directly or wait til the PR lands and then edit the text (none of the changes would require technical review) in master (ie. gh-pages).  @jannyHou LMK which you'd prefer.

**Review by crandmck — CHANGES_REQUESTED** [review]

File: pages/en/lb4/Creating-components.md
Comment (dhmlau) [inline pages/en/lb4/Creating-components.md]: - `When bind a component to an app` --> When *binding* a component to an app
- typo: propertis  --> properties

File: pages/en/lb4/Creating-components.md
Comment (dhmlau) [inline pages/en/lb4/Creating-components.md]: - typo: stragety --> strategy
- Not sure if i understand what it means..  Is it sth like:
a strategy is to use mixin to extends a class with new properties and methods ?

File: pages/en/lb4/Creating-components.md
Comment (dhmlau) [inline pages/en/lb4/Creating-components.md]: suggested change.. simply say sth like:
For more details, see [Mixin](Mixin.htm).

File: pages/en/lb4/Mixin.md
Comment (dhmlau) [inline pages/en/lb4/Mixin.md]: perhaps: "returns a new class extends the base class.." --> "returns a new class extending the base class..."

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: hi @crandmck I think now this PR is still waiting for technical review, I will let you know when it's done but before merging.

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: typo here in the code snippet, should be "implements"

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: implements

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: missing a space: `async value(): number`

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: missing a space: `value(): any`

**Review by kjdelisle — CHANGES_REQUESTED** [review]
I know you said focus on the technical aspects, but the typos I left comments about are code style errors or outright syntax errors.

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: @dhmlau yep.

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: I would like to have the current version since that document is not a must read for people who already knows JS `mixin`. A simplified version implies `Mixin.md`contains something additional/different to the JS concept, but it doesn't :)

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: horrible typo :(

File: pages/en/lb4/Using-components.md
Comment (bajtos) [inline pages/en/lb4/Using-components.md:38]: Unrelated whitespace change, could you please revert?

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: This example has an important problem: it's assigning a class constructor to a variable `newClass`. The usual convention is to start class names with a capital letter, e.g. `NewClass`. However, because we are using `let` + assignment instead of `class`, the name `NewClass` will trigger a tslint error about violating naming conventions for variables.

IMO, we should not promote MixinBuilder and tell people to compose mixins directly via regular JS/TS function invocation and class inheritance:

```ts
class NewClass extends FooMixin(BarMixin(baseClass)) {
}
```

Another important difference to consider:

In my proposed approach, the javascript runtime see a correct class name (`NewClass`) and `NewClass.toString()` operator returns `class NewClass...`:

```
> NewClass.name
"NewClass"

> NewClass.toString()
"class NewClass extends Mixin1(Mixin2(BaseClass)) {
        }"
```

Compare it with the result of `MixinBuilder.mix().with()`, which incorrectly sets the constructor name to `BaseClass` and where `newClass.toString()` returns a fully mixed class definition that makes it difficult to see what have been added by what mixin:

```
> newClass.name
"BaseClass"

> newClass.toString()
"class extends superClass {
        constructor() {
            super(...arguments);
            this.mixinProp2 = 'mixinProp2';
        }
        static staticMixinMethod2() {
            return 'mixin2.static';
        }
        mixinMethod2() {
            return 'mixin2';
        }
    }"
```

I am using the following unit-test code to verify my assumptions and produce sample output: https://github.com/strongloop/loopback-next/blob/568c6e2e7242949e1e5ece040bc3a54d83ff1411/packages/repository/test/unit/mixin/mixin.ts#L53

cc @raymondfeng as the author of MixinBuilder

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: I am not sure if I understand this question. Interfaces are compile-time things that are used by TypeScript compiler to verify type correctness. They are omitted from the transpiled javascript output and they are not available at runtime.

I think it's possible to declare that an interface extends a class, which means that the interface must have all prototype members (properties & methods) of the class being extended. However, no implementation details (no JS code) from the class is inherited this way, it's all about compile-time shape of objects (the type system) only.

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: I think the usual convention for creating mixins is to use anonymous class:

```ts
return class extends baseClass {
  // ...
};
```

I think V8 is smart enough to infer a better class name in such case, most likely use baseClass's name. E.g. `timeStampMixin(UserController)` returns a class called `UserController` and `timeStampMixin(ProductController)` returns a class called `ProductController`, which is much better than having both calls return classes with the same name `newClass`.

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: Ditto, just `return class extends baseClass`

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: As I commented above, for cases where the list of mixins is known at compile time, we should not be using `MixinBuilder`.

File: pages/en/lb4/Mixin.md
Comment (bajtos) [inline pages/en/lb4/Mixin.md]: It's a good practice to give methods names that start with a verb, i.e. `log(str: string)` in this case.

**Review by bajtos — CHANGES_REQUESTED** [review]
Please see my comments below.

In addition to them, please include reference to the following materials explaining ES2015 and TypeScript mixins in more details:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Mix-ins
- http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/
- https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html

File: pages/en/lb4/Mixin.md
Comment (b-admike) [reply pages/en/lb4/Mixin.md]: >we should not promote MixinBuilder and tell people to compose mixins directly via regular JS/TS function invocation and class inheritance:

I wonder if one of the objectives of our MixinBuilder is to compose mixins for this use case. If that's the case, IMO shouldn't the constructor name and class def be preserved by MixinBuilder? (unless there is a reason I'm not aware of)

File: pages/en/lb4/Creating-components.md
Comment (b-admike) [inline pages/en/lb4/Creating-components.md]: Here I think you have some options. You could:
 - Give an overview of `RepositoryMixin` and point to its implementation
 - Give a brief overview of `RepositoryMixin` with code snippets like you said, and point to the source code at the end

I think the first option might suffice, given that Mixins are described in detail in their own page. IMO, the main points this section should cover would be:
- How component developers can leverage mixins
- Guideline on how they should use mixins 
- other considerations that they should take into account

File: pages/en/lb4/Mixin.md
Comment (b-admike) [inline pages/en/lb4/Mixin.md]: I wonder if we should walk readers through this section first before `Mixin Builder` section. IMO I see it as introducing the concept first, then showing them how we make their lives easier with our `MixinBuilder` :)

Comment (jannyHou) [reply pages/en/lb4/Mixin.md]: I also think of the sequence when wrote it. Eventually decided to introduce `MixinBuilder` first:

`MixinBuilder.mix(baseClass).with(FooMixin, BarMixin)` is a very straight forward syntax to give a new  mixin learner a big picture of what is the relation between class and mixin, then they go into section#define-mixin to dig more into implementation and usage. And to tell people already know mixin how `MixinBuilder` provides a prettier syntax to apply mixins.

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: @bajtos good catch...I thought the doc is out-of-date so update it according to our latest tsdoc in `provider.ts`. But didn't think it in the JS compatible direction.

Let me verify it.

File: pages/en/lb4/Mixin.md
Comment (jannyHou) [reply pages/en/lb4/Mixin.md]: I should have removed this, it is a question I am curious to know when reading typescript's class docuemnt:
search "Interfaces Extending Classes" on page https://www.typescriptlang.org/docs/handbook/interfaces.html

`MixinBuilder`'s input is typed to `Class`(we defined in [common-types.ts](https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/common-types.ts))

```ts
constructor(public baseClass: Class<any>) {}
```

So I am wondering would `Interface` work as well.

File: pages/en/lb4/Mixin.md
Comment (jannyHou) [reply pages/en/lb4/Mixin.md]: @raymondfeng Could you take a look of comment https://github.com/strongloop/loopback.io/pull/491/files#r145687471? Thanks!

My understanding of `MixinBuilder` is it provides a prettier syntax to compose mixins to a class, but @bajtos has a good point in comment above.

Comment (raymondfeng) [reply pages/en/lb4/Mixin.md]: I borrowed the idea from http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/

Comment (jannyHou) [reply pages/en/lb4/Mixin.md]: FYI 
> Compare it with the result of MixinBuilder.mix().with(), which incorrectly sets the constructor name to BaseClass and where newClass.toString() returns a fully mixed class definition that makes it difficult to see what have been added by what mixin:

This could be solved by defining `newClass` as 
`newClass = class extends MixinBuilder.mix(BaseClass).with(Mixin1, Mixin2) {};`

Then print out:
```ts
console.log(newClass.name);
// newClass
console.log(newClass.toString());
// class extends _1.MixinBuilder.mix(BaseClass).with(Mixin1, Mixin2) {}
```

---

The problem of `MixinBuilder` from my test is: function `with()` doesn't set its returned type as a class extends the `BaseClass`:
[click to see example test](https://github.com/strongloop/loopback-next/blob/master/packages/repository/test/unit/repository-mixin/repository-mixin.test.ts#L60)
```ts
let AppWithRepoMixin = class extends MixinBuilder.mix(Application).with(RepositoryMixin){};

const myApp = new AppWithRepoMixin();
const boundComponentsBefore = myApp.find('components.*').map(b => b.key);
``` 
fails when compiling, because compiler doesn't know `myApp` is an instance of `Application`, therefore doesn't know it has a prototype function `find()` returning an `Array`. The code calls `map` with `any` instead of `Array` and fails.

This looks like a stop for people wanting to use it...

I am going to defer writing doc for `MixinBuilder` and create an issue for it. I still love the fluent api and would like to dig more :star:

File: pages/en/lb4/Using-components.md
Comment (jannyHou) [reply pages/en/lb4/Using-components.md:38]: weird...I could not see where is the whitespace(difference), I copy paste the original file but diff is still here...

**Comment by jannyHou** [comment]
Apply feedback in https://github.com/strongloop/loopback.io/pull/491/commits/8b43c0f7121676320295eee5863554d175d03ba2:

- rename sidebar title to `Lauguage-related concepts`
- removed document of `MixinBuilder`, details see discussion https://github.com/strongloop/loopback.io/pull/491#discussion_r145687471 and issue https://github.com/strongloop/loopback-next/issues/673 
- complete the creating component doc

**Review by crandmck — APPROVED** [review]

**Review by b-admike — APPROVED** [review]

**Comment by jannyHou** [comment]
@bajtos @kjdelisle feedback applied, PTAL again, thanks!

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: `components`

File: pages/en/lb4/Mixin.md
Comment (kjdelisle) [inline pages/en/lb4/Mixin.md]: `to learn more`.

File: pages/en/lb4/Mixin.md
Comment (kjdelisle) [inline pages/en/lb4/Mixin.md]: `to a record when a controller instance is created`

File: pages/en/lb4/Mixin.md
Comment (kjdelisle) [inline pages/en/lb4/Mixin.md]: `A logger mixin to provide logging tools`

File: pages/en/lb4/Using-components.md
Comment (kjdelisle) [reply pages/en/lb4/Using-components.md:38]: Sounds like a character formatting problem (was this file made on Windows or something? :P)

**Review by kjdelisle — CHANGES_REQUESTED** [review]
Minor nits.

File: pages/en/lb4/Creating-components.md
Comment (jannyHou) [reply pages/en/lb4/Creating-components.md]: @kjdelisle I think it should be `the component's`

whole sentence is:
`When binding a component to an app, you may want to extend the app with the *component's* properties and methods. `

Comment (kjdelisle) [reply pages/en/lb4/Creating-components.md]: I don't see `with the component's` in the existing snippet. ;)
I'm alright with that change as well.

**Comment by jannyHou** [comment]
@kjdelisle thank you :) https://github.com/strongloop/loopback.io/pull/491/commits/8c8890dcf1dd00d14cb4ac26cef74128cc67550a

Really couldn't figure out how `- [Repositories](Repositories.html)` differs from the original file... so I leave it like this :(

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: `If you are not familiar with the mixin concept, see [Mixin](Mixin.htm) to learn more.`

File: pages/en/lb4/Creating-components.md
Comment (kjdelisle) [inline pages/en/lb4/Creating-components.md]: `by using mixins`

**Review by kjdelisle — CHANGES_REQUESTED** [review]
Two more tiny fixes

File: pages/en/lb4/Mixin.md
Comment (bajtos) [reply pages/en/lb4/Mixin.md]: > I am going to defer writing doc for MixinBuilder and create an issue for it. I still love the fluent api and would like to dig more ⭐️

@jannyHou could you please paste a link to that new issue here for posterity?

**Review by bajtos — APPROVED** [review]

File: pages/en/lb4/Mixin.md
Comment (jannyHou) [reply pages/en/lb4/Mixin.md]: @bajtos sure, sorry forgot to link it here: https://github.com/strongloop/loopback-next/issues/673

**Comment by jannyHou** [comment]
@slnode test please

**Review by kjdelisle — APPROVED** [review]

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
@slnode test please

**Pull request merged**
